packages/sdk-core/src/main/kotlin/ly.iterative.itly/core/Itly.kt

```diff
@@ -174,128 +100,109 @@ open class Itly {
-            }
-        }
-
-        return validation;
+        runOnAllPlugins("reset") { it.reset() }
     }
 
     fun flush() {
-        if (this.disabled) {
+        if (disabled) {
             return
         }
 
-        synchronized(enabledPlugins) {
-            enabledPlugins.forEach {
-                try {
-                    it.flush()
-                } catch (e: Exception) {
-                    config.logger.error("$LOG_TAG Error in ${it.id()}.flush(). ${e.message}.")
-                }
-            }
-        }
+        runOnAllPlugins("flush") { it.flush() }
     }
 
     @Synchronized fun shutdown() {
-        if (this.disabled) {
+        if (disabled) {
             return
         }
 
         isShutdown.getAndSet(true)
-        synchronized(enabledPlugins) {
-            enabledPlugins.forEach {
-                try {
-                    it.shutdown()
-                } catch (e: Exception) {
-                    config.logger.error("$LOG_TAG Error in ${it.id()}.shutdown(). ${e.message}.")
-                }
+        runOnAllPlugins("shutdown") { it.shutdown() }
+    }
+
+    private fun runOnAllPlugins(op: String, name: String = "", method: (plugin: Plugin) -> Unit) {
+        config.plugins.forEach {
+            try {
+                method(it)
+            } catch (e: Exception) {
+                config.logger.error("$LOG_TAG Error in ${it.id()}.${op}(${name}). ${e.message}.")
             }
         }
     }
 
-    @Throws(IllegalStateException::class)
-    private fun shouldBeTracked(event: Event): Boolean {
-        var shouldTrack = true;
+    @Throws(IllegalArgumentException::class)
+    private fun validate(event: Event): List<ValidationResponse> {
+        val validationResponses: MutableList<ValidationResponse> = mutableListOf()
+
         if (!config.validation.disabled) {
-            shouldTrack = validate(event).valid;
-            if (config.validation.trackInvalid) {
-                shouldTrack = true;
+            runOnAllPlugins("validate") {
+                val response = it.validate(event)
+                if (response != null) {
+                    validationResponses.add(response)
+                }
             }
         }
-        return shouldTrack
+
+        return validationResponses
+    }
+
+    @Throws(IllegalArgumentException::class)
+    private fun validateAndRunOnAllPlugins(
+        op: String, event: Event,
+        includeContext: Boolean,
+        method: (plugin: Plugin, event: Event) -> Unit,
+        postMethod: (plugin: Plugin, event: Event, validationResponses: List<ValidationResponse>) -> Unit
+    ) {
+        val contextValidationResponses = if (includeContext) validate(context) else listOf()
+        val isContextValid = contextValidationResponses.all { it.valid }
+
+        val eventValidationResponses = validate(event)
+        val isEventValid = eventValidationResponses.all { it.valid }
+
+        val combinedEvent = if (includeContext)
+            Event(
```

Comment:
Is this the best way to combine the props? Also sucks we have to create a brand new event but couldn't think of anything better.

Reply:
Maybe `event.apply { properties = properties + context.properties }`?